fil_fungible_token/src/token/state.rs

```diff
@@ -187,14 +191,53 @@ impl TokenState {
         Ok(new_balance)
     }
 
+    /// Set the balance of the account returning the old balance
+    pub fn set_balance<BS: Blockstore>(
+        &mut self,
+        bs: &BS,
+        owner: ActorID,
+        new_balance: &TokenAmount,
+    ) -> Result<TokenAmount> {
+        // if the new balance is negative, return an error
+        if new_balance.is_negative() {
+            return Err(StateError::NegativeBalance { amount: new_balance.clone(), owner });
+        }
+
+        let mut balance_map = self.get_balance_map(bs)?;
+        let old_balance = match balance_map.get(&owner)? {
+            Some(amount) => amount.clone(),
+            None => TokenAmount::zero(),
+        };
+
+        // if the new balance is zero, remove from balance map
+        if new_balance.is_zero() {
+            balance_map.delete(&owner)?;
+            self.balances = balance_map.flush()?;
+            return Ok(old_balance);
+        }
+
+        // else, set the new balance
+        balance_map.set(owner, new_balance.clone())?;
+        self.balances = balance_map.flush()?;
+        Ok(old_balance)
+    }
+
     /// Retrieve the balance map as a HAMT
-    fn get_balance_map<'bs, BS: Blockstore>(
+    pub fn get_balance_map<'bs, BS: Blockstore>(
         &self,
         bs: &'bs BS,
     ) -> Result<Map<'bs, BS, ActorID, TokenAmount>> {
         Ok(Hamt::load_with_bit_width(&self.balances, bs, HAMT_BIT_WIDTH)?)
     }
 
+    /// Retrieve the number of token holders
+    ///
+    /// This involves iterating through the entire HAMT
+    pub fn get_number_token_holders<BS: Blockstore>(&self, bs: &BS) -> Result<usize> {
```

Comment:
`get_balance_count`? `balance_count`? `count_balances` (sounds slow, like it is)?